sync/handlers.py

```diff
@@ -195,27 +203,34 @@ def __call__(self, git_gecko, git_wpt, body):
         # the decision task is complete. This allows us to determine if a
         # try push should have the expected wpt tasks just by checking if
         # this is set
-        if result not in (tc.SUCCESS, tc.FAIL, tc.EXCEPTION):
-            logger.info("Decision task is not yet complete")
+        if state != "completed" or result == "superseded":
+            logger.info("Decision task is not yet complete, status %s" % result)
             return
 
-        logger.info("Setting taskgroup id for try push %r to %s" % (try_push, task_id))
-        try_push.taskgroup_id = task_id
-
-        failed_builds = len(try_push.failed_builds())
-        if result != tc.SUCCESS or failed_builds:
-            try_push.status = "complete"
-            try_push.infra_fail = True
+        # If we retrigger, we create a new taskgroup, with id equal to the new task_id.
+        # But the retriggered decision task itself is still in the original taskgroup
+        # In order to count the number of retriggers we have, we always retrigger the
+        # first scheduled decision task, so that all decision tasks are in a single taskgroup
```

Comment:
That's not what we're actually doing here,  since `task_id` doesn't always represent the original decision task. We're actually retriggering the most recent retriggered decision task, which is fine. 

Unfortunately, `body` doesn't give us the task group id, so I think we need to save some state. Maybe a retries count in TryPush, rather than saving the original taskgroup id? In that case, we can also modify `retrigger` to make the taskgroup id match the task id whenever we're retry a decision task, thus maintaining the convention that decision task id and taskgroup id always match.

Reply:
I mean, in `retrigger` we can set the payload to have whatever taskgroup id we like...

Reply:
So I think the comment could be wrong, but the logic is "right", depending on what properties we want to preserve.

If we retrigger the `task_id` we are given then the `taskGroupId` will be the same in the retriggered task. For the first retrigger, that will be the same as the original `taskGroupId` and by induction it will continue to be the same for all subsequent retriggers. That's what's implemented here and it does make the logic simple because we have a single taskgroup with all the decision tasks, so it's obvious how many times we have retriggered, just by counting.

It is also possible to set the `taskGroupId` to be equal to the `task_id`, in which case we preserve the property that they are identical for the decision task. *Not* doing that, I think, only has the consequence that the decision task itself ends up in a different task group from the other jobs. However if we do that there's no easy way to look up all the retriggers (I guess we could use treeherder's API if we really wanted), so we probably have to store a retrigger count on the `TryPush` object. 

I think the implemented approah has the benefit of being the simplest thing that could possibly work, but if you want to preserve the stronger invariant on decision task's ids matching, we need to do something more complex.

Reply:
@jgraham I agree with most of that, except that the logic is still wrong. :) The handler is not receiving a task's taskgroup id in `body`, so although all the retriggered decision tasks do belong to the same taskgroup, the following code is not actually getting that taskgroup:

```
# task_id is the id of the retriggered decision task, not the original taskgroup
taskgroup = tc.TaskGroup(task_id)
if len(taskgroup.tasks.view(lambda x: x["metadata"]["name"] == "Gecko Decision Task")) > 5:
    ...
```
The response we'll get is something like https://queue.taskcluster.net/v1/task-group/BpxdADiRTnu67LXP0OYdow/list -- it won't include any decision tasks at all because it's the list of tasks scheduled as a result of the retriggered decision task.

Now that I think about it, you can make an extra request to get more info about the retriggered task: e.g. `tc.get_task(task_id).get("taskGroupId")` and then use that make your `tc.TaskGroup`.